### ISSUE TITLE
Document the differences between t.a.service.IService.disownServiceParent and t.a.service.IServiceCollection.removeService

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11376]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#671|
|Type|defect|
|Created|2004-08-14 02:57:43Z|
|Branch|https://github.com/twisted/twisted/tree/doc-diff-service-671|

<details><summary>Searchable metadata</summary>

```
trac-id__671 671
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_doc_diff_service_671 branches/doc-diff-service-671
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1092452263000000 1092452263000000
changetime__1226850028000000 1226850028000000
version__ 
owner__ 
cc__exarkun cc__moshez cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
These methods do similar things, and it is not clear which is the correct way
to remove a service from a service collection.  I believe disownServiceParent is
the correct way.  This should be documented clearly.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement commented|
|-|-|

```
#!html
<pre>
Thanks. As far as I can tell, all of the Service architecture needs far better docs.

Will likely get to this in a couple of weeks.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement commented|
|-|-|

```
#!html
<pre>
Is removeService deprecated or does it have some internal use?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
&lt;itamar> disownServiceParent is symmetric with setServiceParent in that it sets
the `parent' attribute on the child, so you probably want to use that and
`removeService' is the hook in the collection that subclasses can override, but
usually you don't call yourself.


It sounds like the code should stay the same and the docs should say something
like the above.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement set owner to edsuom|
|-|-|

Re-assigning doc bugs to edsuom.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra commented|
|-|-|

(In [24555]) Branching to 'doc-diff-service-671'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

Mostly (but not entirely) trivial points:

  * use `C{}` instead of ````'`
  * ''symmetric'' is an adjective; the text should use ''symmetrically'' instead because it is modifying a verb (''used'') and so the adverb form is required.
  * it doesn't make sense to talk about a ''subclass'' in the docstring for a method on an interface.  it should talk about what an ''implementation'' would do, instead.
  * ''control'' is non-specific; I think the docstring needs to talk about how an `IServiceCollection` needs to implement `removeService` to take a service out of the collection.  It would also help if it explained that (perhaps by convention, only?  I'm still not completely clear on the intent of the dual API is here) `removeService` is '''only''' responsible for cleaning up state on the `IServiceCollection` implementation.  This is different from `IService.disownServiceParent` in that `disownServiceParent` implementations are required (or are they?  does convention dictate this?  I dunno) to call the `IServiceCollection.removeService` in addition to cleaning up any state on the `IService` itself (such as the `parent` attribute).

Perhaps it would be best to dispense with the text from itamar's suggestion entirely and just say something like ''Use this API to remove an IService from an IServiceCollection'' in the docstring for `IService.disownServiceParent` and something like ''Only implementations of `IService.disownServiceParent` should use this method.'' in the docstring of `IServiceCollection.removeService`.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Fixes in r25320, up for review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set owner to @thijstriemstra|
|-|-|

Great! Small cleanups:
 * remove trailing whitespaces of the file
 * I like to see a bit of indentation after epydoc fields, like this:
```
@return: a L{Deferred} which is triggered when the service has
    finished shutting down. If shutting down is immediate, a value can
    be returned (usually, C{None}).
```

Then you can merge.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra set status to `assigned`|
|-|-|

Replying to [therve](#note_12):
> Great! Small cleanups:
>  * remove trailing whitespaces of the file
>  * I like to see a bit of indentation after epydoc fields, like this:
> 
> @return: a L{Deferred} which is triggered when the service has
>     finished shutting down. If shutting down is immediate, a value can
>     be returned (usually, C{None}).
> 

Addressed these points in r25376

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_14"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [25377]) Merge doc-diff-service-671: Document the differences between t.a.service.IService.disownServiceParent and t.a.service.IServiceCollection.removeService.

Author: thijs
Reviewer: exarkun, therve
Fixes: [#11376](https://github.com/twisted/twisted/issues/11376)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
